config/memory_persister.go

```diff
@@ -4,56 +4,67 @@
 package config
 
 import (
-	"bytes"
-	"io"
-	"io/ioutil"
+	pb "github.com/square/quotaservice/protos/config"
+	"sync"
+	"github.com/juju/errors"
 )
 
 type MemoryConfigPersister struct {
 	config  string
-	configs map[string][]byte
+	configs map[string]*pb.ServiceConfig
 	*Notifier
+	*sync.RWMutex
 }
 
 func NewMemoryConfigPersister() ConfigPersister {
 	p := &MemoryConfigPersister{
-		configs:  make(map[string][]byte),
-		Notifier: NewNotifier()}
+		configs:  make(map[string]*pb.ServiceConfig),
+		Notifier: NewNotifier(),
+		RWMutex: &sync.RWMutex{}}
 
 	p.Notify()
 	return p
 }
 
-// PersistAndNotify persists a marshalled configuration passed in.
-func (m *MemoryConfigPersister) PersistAndNotify(marshalledConfig io.Reader) error {
-	b, err := ioutil.ReadAll(marshalledConfig)
-	if err != nil {
-		return err
+// PersistAndNotify persists a configuration passed in.
+func (m *MemoryConfigPersister) PersistAndNotify(oldHash string, cfg *pb.ServiceConfig) error {
+	m.Lock()
+	defer m.Unlock()
+
+	// Optimistic lock failure
+	if m.config != oldHash && oldHash != initialHash {
+		return errors.Errorf("OldHash passed in %v is not the same as persisted old hash %v", oldHash, m.config)
```

Comment:
nit: this isn't really an "error", is it? We almost expect that it could happen. Not sure of the internal implications of logging at level `Error`.

Reply:
This isn't logging; it's creating a new instance of `error` and returning it.

Reply:
But also, removing for now.